[ADD] 프로젝트 모델링 구성 (ADD: VOIDOC project MODELING)

jhwang0801 opened the pull request:
## :: 최근 작업 주제 
- [x] 기능 추가
- [ ] 리뷰 반영
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 
### - VOIDOC 프로젝트 모델링 구성

<img width="1024" alt="스크린샷 2022-06-22 오전 10 52 53" src="https://user-images.githubusercontent.com/75832544/174960863-86df91c2-0e36-4bca-8298-2782545afaef.png">


<br />

## :: 구현 사항 설명
### 1. CREATE users APP
- 유저 관련 정보에 대해서 `users/models.py`에 모델링 완료
- `users.models` ➤ usermanager, user, doctor, working_day, working_time,  hospital, department 테이블 작성
- `AbstractBaseUser`를 사용하기 위하여 `settings.py`에 `AUTH_USER_MODEL` 설정 완료
- `users/admin.py`에 `admin.site.register(User)`설정 완료

### 2. CREATE appointments APP
- 예약 기능 관련 정보에 대해서 `appointments/models.py`에 모델링 완료
- `appointments.models` ➤ appointment, user_appointment, state, appointment_image 테이블 작성
- `user_appointment` 테이블에는 `user_id`, `doctor_id`, `appointment_id` 3가지의 ForeignKey로 구성
 
### 3. 파일 업로드 기능(로컬 서버 이용)
- media 경로 추가
- 모델 생성(FileField)

### 4. Migration & Migrate 완료
- DB 생성 완료 (cofirmed)

jhwang0801 committed:
ADD: VOIDOC project MODELING

- VOIDOC 프로젝트 모델링 구성
- makemigration & migrate 완료
- django.contrib.auth의 AbstractBaseUser model 사용
- settings.py에Auth_USER_MODEL 설정
- users/admin.py에 admin.site.register(User) 설정
- FileField 사용하기 위하여 media 경로 추가

jasonkang14 commented on appointments/models.py:
```diff
@@ -0,0 +1,34 @@
+        db_table = 'appointments'
+
+class UserAppointment(models.Model): 
+    user        = models.ForeignKey('users.User', on_delete=models.CASCADE)
```
유저 테이블과 연관이 있긴  하지만 예약 테이블에서 보기에는 `patient`라는 이름이 더 가독성에 좋을 것 같습니다

jhwang0801 replied:
아! 감사합니다.
가독성 부분에서 환자와 의사를 확실하게 구분 가능해졌습니다.

수정하여 반영하였습니다!!

jasonkang14 commented on users/models.py:
```diff
@@ -0,0 +1,97 @@
+        user = self.create_user(
+            email    = self.normalize_email(email),
+            name     = name,
+            password = password
```
`is_doctor`에 default parameter가 설정되지 않아서 에러가 발생할 것 같습니다

jhwang0801 replied:
리뷰해주신 내용 테스트해보니,
`TypeError: create_superuser() missing 1 required positional argument: 'is_doctor'`
해당 에러 발생하는 것을 확인하였습니다.

아래와 같이 코드 수정하였고, `python manage.py createsuperuser` 통해서 생성되는 것 까지 확인했습니다. 리뷰 감사합니다!
```python
def create_superuser(self, name, email, is_doctor, password):

    user = self.create_user(
        email     = self.normalize_email(email),
        name      = name,
        is_doctor = is_doctor,
        password  = password
    )

    user.is_superuser = True
    user.save(using=self._db)
    return user


class CustomUser(AbstractBaseUser):
    ...
    objects = UserManager()

    USERNAME_FIELD  = 'email'
    REQUIRED_FIELDS = ['name', 'password', 'is_doctor']

jasonkang14 commented on users/models.py:
```diff
@@ -0,0 +1,97 @@
+        return user
+
+class User(AbstractBaseUser):
+    name         = models.CharField(max_length=50)
```
피그마 화면에 보시면 성과 이름을 나누어서 입력하게 되어있습니다. 
`first_name`, `last_name`으로 분리가 필요합니다. 만약 `name`으로 통일하려면 클라이언트와 협의하셔서 `first_name`, `last_name`인풋을 따로 받더라도 서버에 전송할 때는 `name: last_name + first_name`으로 작업하도록 협의가 필요하겠습니다

jhwang0801 replied:
> 피그마 화면에 보시면 성과 이름을 나누어서 입력하게 되어있습니다. `first_name`, `last_name`으로 분리가 필요합니다. 만약 `name`으로 통일하려면 클라이언트와 협의하셔서 `first_name`, `last_name`인풋을 따로 받더라도 서버에 전송할 때는 `name: last_name + first_name`으로 작업하도록 협의가 필요하겠습니다

해당 부분 자세히 설명을 미리 못드려서 죄송합니다.

프론트엔드 분들께서 `first_name`과 `last_name`을 합쳐서 `name`이라는 형태로 보내주실 수 있다고 말씀하셔서 따로 분리하지 않고 `name`으로 작성하였습니다.

jasonkang14 commented on users/models.py:
```diff
@@ -0,0 +1,97 @@
+    def has_module_perms(self, app_label):
+        return True
+
+    @property
```
:+1

jasonkang14 commented on users/models.py:
```diff
@@ -0,0 +1,97 @@
+
+    @property
+    def is_staff(self):
+        return self.is_superuser
```
is_staff권한은 어떤 용도로 사용할 예정인가요?

jhwang0801 replied:
일차적으로 해당 코드를 사용한 이유는,
해당 코드가 없이 마이그레이션을 할 경우 user 객체에 `is_staff` attribute가 없다는 `AttributeError`가 발생하여 찾아보던 와중에, `is_staff` property를 생성해주는 방식으로 해결이 가능하다는 글을 보았고 이후 에러가 해결되었습니다.

추가로, 해당코드는
모든 `superuser`가 `staff`에 해당한다는 의미로 사용하였습니다. (All superusers are staff) 

어떤 용도로 사용하기 위해서 작성한 코드는 아닙니다.
혹시 사용 목적이 없다면 제거해야할 코드일까요?

jhwang0801 replied:
어떤 용도로 사용하기 위해서 작성한 코드는 아닙니다.
혹시 사용 목적이 없다면 제거해야할 코드일까요?

jasonkang14 commented on voidoc/settings.py:
```diff
@@ -136,3 +140,10 @@
 )
+
+# Auth_User_Model
+AUTH_USER_MODEL = 'users.User'
```
재민님이 작성한 `AUTH_USER_MODEL`과 차이가 있는데 두분이 협의하시고 통일해주시기 바랍니다.

jhwang0801 replied:
재민님께서 작성하신 코드 확인하였습니다.

App과 model의 이름이 통일되지 않은 것으로 확인했는데, 해당부분에 대해서 언급해주신 것이 맞을까요?

jhwang0801 replied:
해당 부분 재민님과 방금 협의하여 통일 시키기로 하였습니다. 감사합니다!

jhwang0801 committed:
Modified(1): project MODELING

- 리뷰 반영
- Merge된 재민님 코드를 pull하여 발생한 conflict fixed
- AbstractBaseUser를 사용하기로 협의한 후 발생한 오류(회원가입, 로그인)들은 일단 주석처리, 현재 함께 수정 중

jhwang0801 committed:
Modified(2): project MODELING

- User Model 이름 변경 to CustomUser
- 로그인 회원가입 기능 모두 제거

jasonkang14 commented on appointments/models.py:
```diff
@@ -0,0 +1,34 @@
+    state      = models.ForeignKey('State', on_delete=models.CASCADE)
+    date       = models.DateField()
+    time       = models.TimeField()
+    created_at = models.DateTimeField(auto_now_add=True)
```
설정을 적절히 잘 사용하셨네요 👍

jasonkang14 commented on voidoc/urls.py:
```diff
@@ -1,10 +1,9 @@
-    path("accounts/", include("django.contrib.auth.urls")),
 ]
+
+if settings.DEBUG:
```
분기를 잘하셨네요  굳굳

jasonkang14 reviewed (approved):
LGTM
